File: server/pkg/api/game/tournament.go
package game

import (
	"github.com/google/uuid"
)

type Tournament struct {
	ID             string   `json:"id"`
	CreatorID      string   `json:"creatorId"`
	UserIDs        []string `json:"userIds"`
	ObservateurIDs []string `json:"observerIds"`
	PoolGames      []*Game  `json:"poolGames"`
	Finale         *Game    `json:"finale"`
	HasStarted     bool     `json:"hasStarted"`
	IsOver         bool     `json:"isOver"`
	WinnerID       string   `json:"winnerId"`
	IsPrivate      bool     `json:"isPrivate"`
}

type TournamentGameInfo struct {
	TournamentID string `json:"tournamentId"`
}

func NewTournament(creatorID string, withUserIDs []string, isPrivate bool) *Tournament {
	t := &Tournament{
		ID:        uuid.NewString(),
		CreatorID: creatorID,
		UserIDs:   []string{creatorID},
		PoolGames: make([]*Game, 0, 2),
		IsPrivate: isPrivate,
	}
	t.UserIDs = append(t.UserIDs, withUserIDs...)

	return t
}

func (t *Tournament) PoolGamesWinners() []string {
	winners := make([]string, 0)
	for _, g := range t.PoolGames {
		if g.WinnerID == "" {
			continue
		}
		winners = append(winners, g.WinnerID)
	}

	return winners
}

File: mobile/lib/models/tournament.dart
import 'package:client_leger/models/game.dart';

class Tournament {
  String id;
  String creatorId;
  List<String> userIds;
  List<String> observateurIds;
  List<Game> poolGames;
  Game? finale;
  bool isOver;
  String winnerId;
  bool isPrivate;
  bool? isProtected;

  Tournament(
      {required this.id,
        required this.creatorId,
        required this.userIds,
        required this.observateurIds,
        required this.poolGames,
        required this.finale,
        required this.isOver,
        required this.winnerId,
        required this.isPrivate,
        required this.isProtected
      });

  factory Tournament.fromJson(Map<String, dynamic> json) {
    int userIdsIndex = 0;
    int observateurIdsIndex = 0;
    int poolGamesIndex = 0;
    return Tournament(
        id: json["id"],
        creatorId: json["creatorId"],
        userIds: List<String>.from((json["userIds"] as List).map(
                (userId) => json["userIds"][userIdsIndex++]
        )),
        observateurIds: List<String>.from((json["observateurIds"] != null
            ? json["observateurIds"] as List : []).map(
                (userId) => json["observateurIds"][observateurIdsIndex++]
        )),
        poolGames: List<Game>.from((json["poolGames"] as List).map(
            (poolGame) => json["poolGames"][poolGamesIndex++]
        )),
        finale: json["finale"],
        isOver: json["isOver"],
        winnerId: json["winnerId"],
        isPrivate: json["isPrivate"],
        isProtected: json["isProtected"]
    );
  }

  Map<String, dynamic> toJson() => {
    "id": id,
    "creatorId": creatorId,
    "userIds": userIds,
    "observateurIds": observateurIds,
    "poolGames": poolGames,
    "finale": finale,
    "isOver": isOver,
    "winnerId": winnerId,
    "isPrivate": isPrivate,
    "isProtected": isProtected
  };
}